### ISSUE TITLE
How can I use FIPS 3.0 provider with Openssl-3.1

### ISSUE THREAD

**Issue by Pankul94** [openssl/openssl#20541]
I want to use the latest Openssl-3.1 and am reading the release notes: https://www.openssl.org/blog/blog/2023/03/07/OpenSSL3.1Release/.

Has the FIPS provider been touched in any way that will change its FIPS certification. The release notes state that the FIPS provider has been upgraded to be FIPS 140-3 compliant but the certification will take time. 

Does that mean that the FIPS provider in Openssl-3.1 is not FIPS certified yet? Or does https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/4282 still apply for the FIPS provider in Openssl-3.1?

The release notes also mention:
> Note that it is possible to upgrade to 3.1 to gain the performance improvements below, and still use the 3.0 FIPS provider.

How can I use 3.0 FIPS provider with Openssl-3.1? Also, is there a way to query for the FIPS provider version from openssl to verify that I am using FIPS 3.0 provider?

**Comment by paulidale** [comment]
The **only** FIPS provider that is certified is the 3.0.0 FIPS provider.  You must build version 3.0.0 of OpenSSL to get this FIPS provider.  It must be downloaded, built and installed **exactly** as per the instructions in the security policy.
Once you've done that, you can build OpenSSL 3.1.0 without FIPS and install it.  This can then be configured to use the 3.0 FIPS provider just like any of the 3.0.x versions.

As for the questions:

- The 3.1.0 FIPS provider is not FIPS certified yet.  Don't expect this until sometime next year.
- No the 4282 certificate does not apply to the OpenSSL 3.1 FIPS provider.
- Query the provider version using `openssl list -providers` or programmatically by getting the `"version"` parameter from the provider after it is loaded.
-

**Comment by paulidale** [comment]
You **must** **only** use the FIPS provider from **3.0.0**.  Building it from later source (3.0.8 here) is **not FIPS approved**.

Apart from that significant detail, you have correctly picked up an older "FIPS" provider.  The 3.0.8 version tells you that.
Now use **3.0.0** to become FIPS validated.

**Comment by Pankul94** [comment]
Understood, thanks!

**Comment by fkonvick** [comment]
> You **must** **only** use the FIPS provider from **3.0.0**. Building it from later source (3.0.8 here) is **not FIPS approved**.
> 
> Apart from that significant detail, you have correctly picked up an older "FIPS" provider. The 3.0.8 version tells you that. Now use **3.0.0** to become FIPS validated.

Uh, I thought that 3.0.8 has been approved, too.
https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/4282

**Comment by paulidale** [comment]
Yes, 3.0.8 is approved now.  The validation needs to be updated when the standards change and CVEs are discovered.
